### ISSUE TITLE
xone-dongle driver is not loading

### ISSUE THREAD

**Issue by atimofeev** [NixOS/nixpkgs#308028]
### Describe the bug
I'm trying to make xbox wireless dongle work, but nothing happens when I insert it.
I've tried numerous different configuration options to no avail.

On the other hand, it is loading on ubuntu 22.04 after a manual setup per official xone instructions.

Any help in finding ways to debug this is highly appreciated.
### Steps To Reproduce
1. `hardware.xone.enable = true;`
2. Reboot
3. Insert dongle

dmesg output:
```
[  140.121147] usb 1-3: new high-speed USB device number 8 using xhci_hcd
[  140.249135] usb 1-3: New USB device found, idVendor=045e, idProduct=02e6, bcdDevice= 1.00
[  140.249143] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  140.249146] usb 1-3: Product: XBOX ACC
[  140.249149] usb 1-3: Manufacturer: MediaTek Inc.
[  140.249151] usb 1-3: SerialNumber: 000000000
[  140.480506] usb 1-3: reset high-speed USB device number 8 using xhci_hcd
[  141.762680] usbcore: registered new interface driver xone-dongle
```

### Expected behavior
xone-dongle driver is properly loaded
dmesg output on ubuntu 22.04:
```
[70703.422904] usb 1-4: new high-speed USB device number 11 using xhci_hcd
[70703.573104] usb 1-4: New USB device found, idVendor=045e, idProduct=02e6, bcdDevice= 1.00
[70703.573127] usb 1-4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[70703.573135] usb 1-4: Product: XBOX ACC
[70703.573141] usb 1-4: Manufacturer: MediaTek Inc.
[70703.573146] usb 1-4: SerialNumber: 000000000
[70703.727536] usb 1-4: reset high-speed USB device number 11 using xhci_hcd
[70703.898183] xone-dongle 1-4:1.0: xone_mt76_send_firmware: build=201703281033____
[70704.021915] xone-dongle 1-4:1.0: xone_mt76_init_radio: id=0x7612
[70704.030910] xone-dongle 1-4:1.0: xone_mt76_init_address: address=62:45:b4:e7:a4:ef
[70704.032532] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=1, power=32
[70704.033108] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=6, power=32
[70704.033656] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=11, power=32
[70704.034231] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=36, power=29
[70704.034744] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=40, power=29
[70704.035301] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=44, power=29
[70704.035804] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=48, power=29
[70704.036354] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=149, power=29
[70704.036856] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=153, power=29
[70704.037358] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=157, power=29
[70704.037901] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=161, power=29
[70704.038515] xone-dongle 1-4:1.0: xone_mt76_evaluate_channels: channel=165, power=29
[70704.038670] xone-dongle 1-4:1.0: xone_mt76_init_channels: channel=1
```

### Notify maintainers
@rhysmdnz 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.86, NixOS, 23.11 (Tapir), 23.11.20240417.e402c3e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by atimofeev** [comment]
Resolved by pairing the gamepad with the dongle on windows pc. 
From what I see, pairing is not working properly on xone side. 
Not a nix issue. Closing this.

**Comment by evilwombat** [comment]
@atimofeev Oh, not you too! Looks like you have the same counterfeit wireless dongle as I do. Even your MAC address is identical to mine: `62:45:b4:e7:a4:ef`

TL;DR: Use the firmware file from https://github.com/ChimeraOS/chimeraos/issues/1020#issuecomment-2804675381 and your pairing button should work again.

**Comment by atimofeev** [comment]
@evilwombat, thank you for your findings! I've managed to fix pairing button on my Ubuntu machine thanks to your guide!

Have to reopen this issue, as I can't make this work on my NixOS machine.
I've tried this overlay:

```nix
final: prev: {
  xow_dongle-firmware = prev.xow_dongle-firmware.overrideAttrs (old: {
    version = "2015-12";
    src = prev.fetchurl {
      url =
        "https://catalog.s.download.windowsupdate.com/d/msdownload/update/driver/drvs/2015/12/20810867_1b8b44bdbdc6a6c60a224d3110c8360d95611b26.cab";
      sha256 = "sha256-OPBOIIqveu4W5rWIY8szL4IuAHPsjGQ4m6apR2/+5os=";
    };
  });
}
```

What I'm missing here?

**Comment by evilwombat** [comment]
Sorry, no idea how to use nix, but the (maintained) version of 'xone' how has proper support for loading the correct firmware file based on the dongle version. I suggest installing from master, here: https://github.com/dlundqvist/xone
You will need to re-run the firmware install script as well, from this version.

**Comment by atimofeev** [comment]
Applied this overlay:
```nix
final: prev: {
  linuxPackages_latest = prev.linuxPackages_latest.extend (lpfinal: lpprev: {
    xone = lpprev.xone.overrideAttrs (oldAttrs: {
      version = "0.3-unstable-2025-04-18";
      src = prev.fetchFromGitHub {
        owner = "dlundqvist";
        repo = "xone";
        rev = "c682b0cd4fd56d2d9639b64787034a375535eb4b";
        sha256 = "sha256-QGmrOCiMa/Nrm2ln7aO+QVL3F5nAK2n6qWMVn+VMwcM=";
      };
      patches = [ ];
    });
  });
}
```
Dongle pairing button still won't work for me. Probably something is missing on nix side.
```
Apr 21 19:13:51 milaptop kernel: xone-dongle 1-3:1.0: Direct firmware load for xow_dongle_045e_02e6.bin failed with error -2
Apr 21 19:13:51 milaptop kernel: xone-dongle 1-3:1.0: xone_mt76_init_address: MAC address 62:45:b4:e7:a4:ef looks suspicious. Counterfeit adapter? That may be fine, but consider passing override_mac= to deconflict with others nearby
```

**Comment by evilwombat** [comment]
You need to rerun the firmware download script, which will pull in the
firmware file for your specific model of dongle. Driver isn't finding this
file and is falling back to the default firmware. That is why the pairing
button isn't working.

**Comment by atimofeev** [comment]
Updated overlay installPhase:
```nix
final: prev: {
  xow_dongle-firmware = prev.xow_dongle-firmware.overrideAttrs (old: {
    version = "2015-12";
    src = prev.fetchurl {
      url =
        "https://catalog.s.download.windowsupdate.com/d/msdownload/update/driver/drvs/2015/12/20810867_1b8b44bdbdc6a6c60a224d3110c8360d95611b26.cab";
      sha256 = "sha256-OPBOIIqveu4W5rWIY8szL4IuAHPsjGQ4m6apR2/+5os=";
    };
    installPhase = ''
      install -Dm644 FW_ACC_00U.bin ${
        placeholder "out"
      }/lib/firmware/xow_dongle_045e_02e6.bin
    '';
  });
}

```
Now the firmware load error is gone, but the pairing button is still non-functional. 

However, when I manually place firmware files into `/lib/firmware/` according to [xone firmware download script](https://github.com/dlundqvist/xone/blob/ffe1251c7c518574bf409f7e915fab8c532544c0/install/firmware.sh), it works as expected.

Dear `xone` package maintainers, could you please assist?
@rhysmdnz 
@fxzzi

**Comment by evilwombat** [comment]
Not sure what's going on with Nix, but the new xone driver expects both firmware files to be present. The original xone_dongle.bin will be used for all standard dongles and the 045e_02e6.bin version will be used for the 02e6 hardware variant. Since the user could plug in a different dongle at any time, both firmware files need to be present in /lib, available for to the driver to choose. The driver already knows to select between them.

**Comment by fxzzi** [comment]
> Dear `xone` package maintainers, could you please assist?
> ...
> [@fxzzi](https://github.com/fxzzi)

hey there!!

I've made a PR over at #400938 , would you be happy to test and leave your results there? I don't have one of these dongles to test and therefore its a little hard for me to know its working. It should be placed in the correct place according to your little overlay `$out/lib/firmware/xow_dongle_045e_02e6.bin`

hopefully it works :)

**Comment by atimofeev** [comment]
I can confirm that #400938 works as expected with both genuine and counterfeit dongles! 🔥

**Comment by c0d3m0nky** [comment]
> [ pkgs.xow_dongle-firmware-045e_02e6 ];

Not working for me, but not 100% sure I applied this right

```
> dmesg
[  +0.073671] mt76x2u 5-4:1.0: Direct firmware load for mt7662_rom_patch.bin failed with error -2
```

My config
```Nix
nixpkgs.overlays = [ 
    (final: prev: {
      xow_dongle-firmware-045e_02e6 = prev.xow_dongle-firmware.overrideAttrs
        (old: rec {
          version = "045e_02e6";
          src = prev.fetchurl {
            url =
              "https://catalog.s.download.windowsupdate.com/d/msdownload/update/driver/drvs/2015/12/20810869_8ce2975a7fbaa06bcfb0d8762a6275a1cf7c1dd3.cab";
            sha256 = "sha256-5jiKJ6dXVpIN5zryRo461V16/vWavDoLUICU4JHRnwg=";
          };
          unpackCmd = ''
            cabextract -F FW_ACC_00U.bin ${src}
          '';
          installPhase = ''
            install -Dm644 FW_ACC_00U.bin ${
              placeholder "out"
            }/lib/firmware/xow_dongle_045e_02e6.bin
          '';
        });
      })
  ]; 
   
hardware.firmware = [ pkgs.xow_dongle-firmware-045e_02e6 ];
```

**Comment by evilwombat** [comment]
That looks like the wrong driver. The Xbox dongle uses the same WiFi chip
as some USB WiFi adapters, and it looks like your system is trying to use a
generic mediatek wireless driver instead of using xone.
